rollup/da_syncer/da/calldata_blob_source.go

```diff
@@ -105,58 +130,139 @@ func (ds *CalldataBlobSource) processRollupEventsToDA(rollupEvents l1.RollupEven
+	for i, commitEvent := range commitEvents {
+		// sanity check commit events from batches submitted in the same L1 transaction
+		if commitEvent.TxHash() != firstCommitEvent.TxHash() {
+			return nil, fmt.Errorf("commit events have different tx hashes, batch index: %d, tx: %s - batch index: %d, tx: %s", firstCommitEvent.BatchIndex().Uint64(), firstCommitEvent.TxHash().Hex(), commitEvent.BatchIndex().Uint64(), commitEvent.TxHash().Hex())
+		}
+		if commitEvent.BlockNumber() != firstCommitEvent.BlockNumber() {
+			return nil, fmt.Errorf("commit events have different block numbers, batch index: %d, block number: %d - batch index: %d, block number: %d", firstCommitEvent.BatchIndex().Uint64(), firstCommitEvent.BlockNumber(), commitEvent.BatchIndex().Uint64(), commitEvent.BlockNumber())
+		}
+		if commitEvent.BlockHash() != firstCommitEvent.BlockHash() {
+			return nil, fmt.Errorf("commit events have different block hashes, batch index: %d, hash: %s - batch index: %d, hash: %s", firstCommitEvent.BatchIndex().Uint64(), firstCommitEvent.BlockHash().Hex(), commitEvent.BatchIndex().Uint64(), commitEvent.BlockHash().Hex())
+		}
+		if previousEvent != nil && commitEvent.BatchIndex().Uint64() != previousEvent.BatchIndex().Uint64()+1 {
+			return nil, fmt.Errorf("commit events are not in sequence, batch index: %d, hash: %s - previous batch index: %d, hash: %s", commitEvent.BatchIndex().Uint64(), commitEvent.BatchHash().Hex(), previousEvent.BatchIndex().Uint64(), previousEvent.BatchHash().Hex())
+		}
+
+		switch codec.Version() {
+		case 0:
+			if entry, err = NewCommitBatchDAV0(ds.db, codec, commitEvent, args.ParentBatchHeader, args.Chunks, args.SkippedL1MessageBitmap); err != nil {
+				return nil, fmt.Errorf("failed to decode DA, batch index: %d, err: %w", commitEvent.BatchIndex().Uint64(), err)
+			}
+		case 1, 2, 3, 4, 5, 6:
+			if entry, err = NewCommitBatchDAV1(ds.ctx, ds.db, ds.blobClient, codec, commitEvent, args.ParentBatchHeader, args.Chunks, args.SkippedL1MessageBitmap, args.BlobHashes, blockHeader.Time); err != nil {
+				return nil, fmt.Errorf("failed to decode DA, batch index: %d, err: %w", commitEvent.BatchIndex().Uint64(), err)
+			}
+		case 7:
+			if i >= len(args.BlobHashes) {
+				return nil, fmt.Errorf("not enough blob hashes for commit transaction: %s, index in tx: %d, batch index: %d, hash: %s", firstCommitEvent.TxHash(), i, commitEvent.BatchIndex().Uint64(), commitEvent.BatchHash().Hex())
+			}
+			blobHash := args.BlobHashes[i]
+
+			var parentBatchHash common.Hash
+			if previousEvent == nil {
+				parentBatchHash = common.BytesToHash(args.ParentBatchHeader)
+			} else {
+				parentBatchHash = previousEvent.BatchHash()
+			}
+
+			if entry, err = NewCommitBatchDAV7(ds.ctx, ds.db, ds.blobClient, codec, commitEvent, blobHash, parentBatchHash, blockHeader.Time); err != nil {
+				return nil, fmt.Errorf("failed to decode DA, batch index: %d, err: %w", commitEvent.BatchIndex().Uint64(), err)
+			}
+		default:
```

Comment:
In case of a zk bug, we'd probably register a new on-chain verifier for version `8` (without changing the batch format). It's nicer if there are no other contract and l2geth changes necessary, so I think we can change `case 7` to `default` here.